### ISSUE TITLE
RouterLink causes the sidebar nav area to redraw on every click

### ISSUE THREAD

**Issue by bhaidar** [angular/angular#24799]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[X] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
The application I am currently working on requires a sidebar navigation menu and a content area (router-outlet). When I click on any link on the sidebar nav (RouterLink), the sidebar is being redrawn together with the content area.

## Expected behavior
The sidebar navigation area shouldn't be refreshed when a RouterLink executes.

## Minimal reproduction of the problem with instructions
`navigateTo(childMenuItem: SideBarMenuItem): (string | number)[] {
    let result: Array<string> = [];
    let defaultUrl = '/dynamic';

    if (childMenuItem && !childMenuItem.multicard && childMenuItem.content[0] && childMenuItem.content[0].typeAlias) {
        const url = `./${childMenuItem.content[0].typeAlias}`;
        result.push(url);
    } else {
        result.push(defaultUrl);
    }

    return result;
}`

That's how I am generating my links.
And that's how I am using it:

` <a [routerLink]="navigateTo(grandChildMenuItem)" class="m-menu__link">`

## Environment

<pre><code>
Angular version: 5.2.5

Browser:
- [X] Chrome (desktop) version 67
 
I've already posted about this issue on Stackoverflow: https://stackoverflow.com/questions/51210982/routerlink-causes-the-entire-page-to-re-draw-itself

Thanks
Bilal

**Comment by ponkys** [comment]
When you do 
```
<a [routerLink]="navigateTo(grandChildMenuItem)" class="m-menu__link">
```

The template checks this function at every doCheck cycle (I'm not sure but that's what I see in the logs). This affects performance considerably and causes that specific area to be redrawn all the time. You should use another approach if you want to avoid this. 

You can generate the URL (ins this case ````./${childMenuItem.content[0].typeAlias}`;```) with the link before you render the template and then just reflect it. You can call the generateURL() method everytime data changes as well.

**Comment by bhaidar** [comment]
Hi @ponkys 
I need to provide more information that I have found right now.

In the app, the AppComponent is something as:
`
<header><header>
<router-outlet></router-outlet>
`
The MainComponent is defined a child in the App.Routes to lazily load a module.
In the lazily loaded module, I define a parent/child relationship as follows:
`
{
        path: '',
        component: AsideLeftDefaultComponent,
        children: [
            {
                path: '',
                component: MainComponent
            }
        ]
    }
`
Where, AsideLeftDefaultComponent is:
`
<app-aside></app-side>
<router-outlet></router-outlet>
`
So basically, what I found out is that when clicking a link inside <app-aside> component, the router-outlet in the AppComponent is being refreshed entirely rather than the router-outlet in the AsideLeftDefaultComponent.

What is needed is to refresh router-outlet of the AsideLeftDefaultComponent rather than refreshing the AppComponent router-outlet.

Thanks
Bilal

**Comment by jotatoledo** [comment]
As already explained, the dynamically generated router link array is calculated on every cd cycle. 
The way you generate this array is independent of the component state as it seems, so the best that you could do would be to capsulate the logic in a pure pipe.

**Comment by bhaidar** [comment]
@jotatoledo Can you illustrate more on "pure pipe"? What should be done, any example?

**Comment by kdubious** [comment]
@bhaidar Are you building your menu items in an *ngFor? Your routerLink `<a [routerLink]="navigateTo(grandChildMenuItem)" class="m-menu__link">` is setting the value to a function call. Can you find a way to work that out in your loop? Perhaps create that function (`navigateTo`) as a property of the objects you loop over to build your `<a>` tags.

If the menu keeps changing, then try to make the menu items an `Observable` / `BehaviorSubject` / etc. and use an `async pipe` in your loop.
https://stackoverflow.com/questions/33747069/problems-using-async-pipe-with-ngfor-in-angular2

**Comment by jasonaden** [comment]
Hello, we reviewed this issue and determined that it doesn't fall into the bug report or feature request category. This issue tracker is not suitable for support requests, please repost your issue on [StackOverflow](http://stackoverflow.com/) using tag `angular`.

If you are wondering why we don't resolve support issues via the issue tracker, please [check out this explanation](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-got-a-question-or-problem).

<hr>

Thanks for the community responses here @kdubious, @ponkys and @jotatoledo!
